docs/proposals/generic-schema-batching-proposal.md

```diff
@@ -0,0 +1,987 @@
+**Key Responsibilities**:
+- Fetch schema metadata from blockchain via existing RPC methods (`getSchema`, `getSchemaPayload`)
+- Maintain two-way caching: `schemaId ↔ SchemaDefinition` and `schemaName ↔ schemaId`
+- Determine if a schema is batchable by inspecting payload location and model type
+- Pre-load known DSNP schemas at startup for backward compatibility
+- Provide cache invalidation for testing/admin purposes
+
+**Schema Definition Interface**:
+```typescript
+interface SchemaDefinition {
+  schemaId: number;                    // 1-65536
+  namespace: string;                    // e.g., "dsnp", "atproto"
+  descriptor: string;                   // e.g., "broadcast", "post"
+  version: string;                      // e.g., "v2"
+  fullName: string;                     // "namespace.descriptor@version"
+  payloadLocation: 'IPFS' | 'OnChain'; // Where content is stored
+  settings: {
+    model: 'Parquet' | 'AvroBinary';   // Batch file format
+  };
+  batchable: boolean;                   // Auto-detected
+  fetchedAt: number;                    // For TTL expiry
+}
+```
+
+**Caching Strategy** (Three-Tier):
+```
+Request → In-Memory Map (instant)
+            ↓ miss
+          Redis Cache (~5ms)
+            ↓ miss
+          Blockchain RPC (~50ms)
+            ↓
+          Cache both layers
+```
+
+**Batchability Detection Logic**:
+- Schema is batchable if: `payloadLocation === 'IPFS'` AND `model === 'Parquet' | 'AvroBinary'`
+- All DSNP content schemas are batchable
+- On-chain schemas are not batchable (published directly)
+
+**Public API**:
+- `getSchema(schemaId): Promise<SchemaDefinition>` - Main lookup method
+- `getSchemaIdByName(fullName): Promise<number>` - Reverse lookup
+- `isBatchable(schemaId): boolean` - Quick batchability check
+- `invalidateSchema(schemaId): Promise<void>` - Admin cache clear
+
+#### 2. Generic Content DTO
+
+**Purpose**: Accept content for any schema without type-specific constraints.
+
+**Location**: `libs/types/src/dtos/content-publishing/generic-content.dto.ts`
+
+**Structure**:
+```typescript
+GenericContentDto {
+  schemaId: number;           // 1-65536, validated
+  content: Record<string, any>; // Pre-formatted per schema
+  referenceId?: string;       // Optional tracking ID
+}
+
+GenericContentResponseDto {
+  referenceId: string;        // For tracking
+  schemaId: number;           // Confirms schema
+  batched: boolean;           // Whether batched or direct
+}
+```
+
+**Key Point**: Content must be **pre-formatted** to match the target schema's structure. The Gateway does no transformation in Solution 1.
+
+#### 3. Generic API Endpoint
+
+**Location**: `apps/content-publishing-api/src/controllers/v1/content.controller.v1.ts`
+
+**New Endpoint**: `POST /v1/content/:msaId/generic`
+
+**Flow**:
+```
+1. Receive request with schemaId + content
+2. Validate schema exists via SchemaRegistry
+3. Check if schema is batchable
+4. If batchable → queue for batching (with schema-specific queue name)
```

Comment:
We might want to consider a different batch/queuing model. One Bull queue per batching schema is not very scalable.